platforms/chibios/drivers/serial_protocol.c

```diff
@@ -0,0 +1,164 @@
+// Copyright 2022 Stefan Kerkmann
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "quantum.h"
+#include "serial.h"
+#include "serial_protocol.h"
+#include "printf.h"
+#include "synchronization_util.h"
+
+#include <ch.h>
+
+static const uint8_t HANDSHAKE_MAGIC = 7U;
+
+static inline bool initiate_transaction(uint8_t sstd_index);
+static inline bool react_to_transactions(void);
+
+/**
+ * @brief This thread runs on the slave and responds to transactions initiated
+ * by the master.
+ */
+static THD_WORKING_AREA(waSlaveThread, 1024);
+static THD_FUNCTION(SlaveThread, arg) {
+    (void)arg;
+    chRegSetThreadName("usart_tx_rx");
+
+    while (true) {
+        if (unlikely(!react_to_transactions())) {
+            /* Clear the receive queue, to start with a clean slate.
+             * Parts of failed transactions or spurious bytes could still be in it. */
+            driver_clear();
+        }
+        split_shared_memory_unlock();
+    }
+}
+
+/**
+ * @brief Slave specific initializations.
+ */
+void soft_serial_target_init(void) {
+    driver_slave_init();
+
+    /* Start transport thread. */
+    chThdCreateStatic(waSlaveThread, sizeof(waSlaveThread), HIGHPRIO, SlaveThread, NULL);
+}
+
+/**
+ * @brief Master specific initializations.
+ */
+void soft_serial_initiator_init(void) {
+    driver_master_init();
+}
+
+/**
+ * @brief React to transactions started by the master.
+ */
+static inline bool react_to_transactions(void) {
+    /* Wait until there is a transaction for us. */
+    uint8_t sstd_index = 0;
+    receive_blocking(&sstd_index, sizeof(sstd_index));
```

Comment:
`receive_blocking` should probably be checked for failure in order to abort `react_to_transactions` cleanly. Of course, for USART this practically doesn't happen, but then again, this serial_protocol overhaul is designed to allow for easy implementation of other protocols, too. And those might be more likely to fail on `receive_blocking`.
->  https://github.com/preisi/qmk_firmware/commit/6dcc3ec22cd9a19a1092808e46f936b382152eec

Reply:
Done!